### ISSUE TITLE
Should curl_multi_info_read return the same easy handle twice in succession?

### ISSUE THREAD

**Issue by lilydjwg** [curl/curl#3324]
### I did this
My program [nvchecker](https://github.com/lilydjwg/nvchecker) uses curl to send out a lot of concurrent requests. This program uses [Tornado](https://github.com/tornadoweb/tornado) which uses pycurl.

Sometimes `curl_multi_info_read` returns the same easy handle twice in succession, breaking Tornado's expectation. I also didn't find any indications in the curl docs that this could happen and how the caller should react.

There are more details in https://github.com/tornadoweb/tornado/issues/2542.

Sorry I don't have a simpler reproduction case.

### I expected the following
`curl_multi_info_read` returns each easy handle once before it has been removed by `curl_multi_remove_handle` then added again by `curl_multi_add_handle`.

### curl/libcurl version

```
curl 7.62.0 (x86_64-pc-linux-gnu) libcurl/7.62.0 OpenSSL/1.1.1 zlib/1.2.11 libidn2/2.0.5 libpsl/0.20.2 (+libidn2/2.0.4) libssh2/1.8.0 nghttp2/1.34.0
Release-Date: 2018-10-31
Protocols: dict file ftp ftps gopher http https imap imaps pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: AsynchDNS IDN IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz TLS-SRP HTTP2 UnixSockets HTTPS-proxy PSL 
```

This doesn't happen with 7.61.1.

### operating system

Arch Linux x86_64.

**Comment by bagder** [comment]
This sounds like a potentially nasty bug and I can't explain how this could happen. But I also don't even know where to begin on how to reproduce this based on what's been presented here.

**Comment by lilydjwg** [comment]
Could you try to send 22 requests to a PyPI URL (e.g. https://pypi.python.org/pypi/tornado/json) with 20 easy handles and see if the issue happens? I'm sorry I don't feel good about writing C code using libcurl (I have never used the libcurl API before).

**Comment by bagder** [comment]
I wrote up [3324.c](https://gist.github.com/bagder/18a520507f932e0f8b128a07b42c60a8) in an attempt to reproduce according to the instructions here. It runs just fine for me...

**Comment by bagder** [comment]
The work-around mentions redirects being part of the issue. My example code doesn't follow redirects. Should it? Should it get some other URL to trigger (more) redirects?

**Comment by lilydjwg** [comment]
Yes, it should follow redirects. I've tried adding that, along with HTTP/2 support, but still can't reproduce it. I'll try to reproduce harder in coming days. Thanks for your code.

**Comment by bdarnell** [comment]
@lilydjwg posted a shorter reproduction program in https://github.com/tornadoweb/tornado/issues/2542#issuecomment-448963191

It's going through [tornado's curl wrapper](https://github.com/tornadoweb/tornado/blob/master/tornado/curl_httpclient.py); the FOLLOW_LOCATION option is the only thing that looks relevant to me that is different between this python case and your 3324.c. We also call `curl_multi_socket_all()` once a second to work around bugs in older versions of libcurl where events could get dropped on the floor without this.

**Comment by lilydjwg** [comment]
I've already tried the FOLLOW_LOCATION option and using different URLs for each request in C. Unfortunately it doesn't reproduce. Maybe to trigger this we need to use an event loop as tornado does, instead of simply using `curl_multi_fdset` and `curl_multi_perform`.

**Comment by bagder** [comment]
If you still experience this problem, I would urge you to work harder on modifying the C example so that it can reproduce. As long as you can't make that code trigger this issue, I will presume this problem is caused by something on top of libcurl and will close this soon.
